plugin/completion.py

```diff
@@ -41,81 +41,68 @@
+            self.view.run_command("insert_snippet", {"contents": insert_text})
+        else:
+            self.view.run_command("insert", {"characters": insert_text})
+        self.handle_additional_edits(item)
 
-    def handle_resolve_response(self, response: Optional[dict]) -> None:
-        if response:
-            additional_edits = response.get('additionalTextEdits')
-            if additional_edits:
-                self.apply_additional_edits(additional_edits)
 
-    def apply_additional_edits(self, additional_edits: List[dict]) -> None:
-        edits = list(parse_text_edit(additional_edit) for additional_edit in additional_edits)
-        debug('applying additional edits:', edits)
-        self.view.run_command("lsp_apply_document_edit", {'changes': edits})
-        sublime.status_message('Applied additional edits for completion')
+class LspCompleteTextEditCommand(LspCompleteCommand):
+
+    def run(self, edit: sublime.Edit, **item: Any) -> None:
+        text_edit = item["textEdit"]
+        new_text = text_edit['newText']
+        edit_region = range_to_region(Range.from_lsp(text_edit['range']), self.view)
+        if item.get("insertTextFormat", InsertTextFormat.PlainText) == InsertTextFormat.Snippet:
+            self.handle_text_edit_snippet(edit, edit_region, new_text)
+        else:
+            self.handle_text_edit_plaintext(edit, edit_region, new_text)
+        self.handle_additional_edits(item)
+
+    def translated_regions(self, edit_region: sublime.Region) -> Generator[sublime.Region, None, None]:
+        selection = self.view.sel()
+        primary_cursor_position = selection[0].b
+        for region in reversed(selection):
+            # For each selection region, apply the same removal as for the "primary" region.
+            # To do that, translate, or offset, the LSP edit region into the non-"primary" regions.
+            # The concept of "primary" is our own, and there is no mention of it in the LSP spec.
+            translation = region.b - primary_cursor_position
+            translated_edit_region = sublime.Region(edit_region.a + translation, edit_region.b + translation)
+            yield translated_edit_region
+
+    def handle_text_edit_snippet(self, edit: sublime.Edit, edit_region: sublime.Region, new_text: str) -> None:
+        for region in self.translated_regions(edit_region):
+            self.view.erase(edit, region)
+        self.view.run_command("insert_snippet", {"contents": new_text})
+
+    def handle_text_edit_plaintext(self, edit: sublime.Edit, edit_region: sublime.Region, new_text: str) -> None:
+        for region in self.translated_regions(edit_region):
+            if len(region) == 0:
+                self.view.insert(edit, region.b, new_text)
+            else:
+                self.view.replace(edit, region, new_text)
 
 
 def resolve(completion_list: sublime.CompletionList, items: List[sublime.CompletionItem], flags: int = 0) -> None:
```

Comment:
Only called from one place now. Maybe inline it?

Reply:
I think that you dont need this check `if len(region) == 0:`
```diff
-            if len(region) == 0:
-                self.view.insert(edit, region.b, new_text)
-            else:
-                self.view.replace(edit, region, new_text)
+           self.view.replace(edit, region, new_text)
```
Because if you are do  view.replace(edit, Region(4,4), 'text'), it will behave like an insert.

Reply:
Also maybe a different approach to make the `handle_text_edit_plaintext` look like `handle_text_edit_snippet`?

```diff
-        for region in self.translated_regions(edit_region):
-            if len(region) == 0:
-                self.view.insert(edit, region.b, new_text)
-            else:
-                self.view.replace(edit, region, new_text)

+        for region in self.translated_regions(edit_region):
+            self.view.erase(edit, region)
+        self.view.insert("insert", {"characters": new_text})
```

Reply:
I would suggest inline the `handle_text_edit_snippet` and `handle_text_edit_plaintext` directly in the `run` method so that the LspCompleteTextEditCommand looks like this at the end.
```
class LspCompleteTextEditCommand(LspCompleteCommand):

    def run(self, edit: sublime.Edit, **item: Any) -> None:
        text_edit = item["textEdit"]
        new_text = text_edit['newText']
        edit_region = range_to_region(Range.from_lsp(text_edit['range']), self.view)

        for region in self.translated_regions(edit_region):
            self.view.erase(edit, region)

        if item.get("insertTextFormat", InsertTextFormat.PlainText) == InsertTextFormat.Snippet:
            self.view.run_command("insert_snippet", {"contents": new_text})
        else:
            self.view.run_command("insert", {"characters": new_text})

        self.handle_additional_edits(item)
```

Reply:
> Only called from one place now. Maybe inline it?

Just to be clear, I meant the `resolve` function :)

Reply:
I accidentally removed the `resolve` call. I inlined the handle_text_edit_snippet and handle_text_edit_plaintext functions. I'm not too keen on calling `self.view.erase` for both plaintext and snippets. A `self.view.replace` for plaintext is much faster than calling an ST command.